Implement a subset of the CWL Draft 3 tool format.

jmchilton opened the pull request:
## CWL Support:
- Implemented integer, long, float, double, boolean, and File parameters, and arrays thereof as well `["null", <simple_type>]` union parameters and Any-type parameters. More complex unions of datatypes are stil unsupported (unions of two or more non-null parameters, unions of `["null", Any]`, etc...).
- Draft 3 `CreateFileRequirement`s are supported (see the `test_rename` test case).
- Draft 3 `InlineJavascriptRequirement` are support to define output files (see `test_cat3` test case).
- `EnvVarRequirement`s are supported (see the `test_env_tool1` and `test_env_tool2` test cases).
- Secondary files are supported at least partially, see the `index1` and `showindex1` CWL tools created to verify this as well as the `test_index1` test case.
- Docker integration is only partial (simple docker pull is supported) - so `cat3-tool.cwl` works for example. Full semantics of CWL docker support has yet to be implemented. The remaining work is straight-forward and trackd in the meta-issue galaxyproject#1684.
- Expression tools are supported (see `parseInt-tool` test case).
- Non-File CWL outputs are represented as `expression.json` files. Traditionally Galaxy hasn't supported non-File outputs from tools but CWL Galaxy has work in progress on bringing native Galaxy support for such outputs #27.
## Implementation Notes:
- CWL secondary files are stored in `__secondary_files__` directory in the dataset's extra_files_path directory.
- The tool execution API has been extended to add a `inputs_representation` parameter that can be set to "cwl" now. The `cwl` representation for running tools corresonding to the CWL job json format with {class: "File: path: "/path/to/file"} inputs replaced with {"src": "hda", "id": "<dataset_id>"}. Code for building these requests for CWL job json is available in the test class.
- Since the CWL <-> Galaxy parameter translation may change over time, for instance if Galaxy develops or refines parameter classes - CWL state and CWL state version is tracked in the database and hopefully for reruns, etc... we could update the Galaxy state from an older version to a new one.
- CWL allows output parameters to be either `File` or non-`File` and determined at runtime, so `galaxy.json` is used to dynamically adjust output extension as needed for non-`File` parameters.
## Implementation Description:

The reference implementation Python library (mainly developed by Peter Amstutz - https://github.com/common-workflow-language/common-workflow-language/tree/master/reference) is used to load tool files ending with `.json` or `.cwl` and proxy objects are created to adapt these tools to Galaxy representations. In particular input and output descriptions are loaded from the tool.

When the tool is submitted, a special specialized tool class is used to build a cwltool compatible job description from the supplied Galaxy inputs and the CWL reference implementation is used to generate a CWL reference implementation Job object. A command-line is generated from this Job object.

As a result of this - Galaxy largely does not need to worry about the details of command-line adapters, expressions, etc....

Galaxy writes a description of the CWL job that it can reload to the job working directory. After the process is complete (on the Galaxy compute server, but outside the Docker container) this representation is reloaded and the dynamic outputs are discovered and moved to fixed locations as expected by Galaxy. CWL allows for much more expressive output locations than Galaxy, for better or worse, and this step uses cwltool to adapt CWL to Galaxy outputs.

Currently all `File` outputs are sniffed to determined a Galaxy datatype, CWL draft 3 allows refinement on this and this remains work to be done.

  1) CWL should support EDAM declaration of types and Galaxy should provide a mapping to core datasets to skip sniffing is types are found.
  2) For finer grain control within Galaxy, extensions to CWL should allow setting actual Galaxy output types on outputs. (Distinction between fastq and fastqsanger in Galaxy is very important for instance.)
## Testing:

```
% git clone https://github.com/common-workflow-language/galaxy.git
% git checkout cwl
% cd galaxy
% virtualenv .venv
% . .venv/bin/activate
% pip install cwltool
```

Start Galaxy.

```
% GALAXY_RUN_WITH_TEST_TOOLS=1 run.sh --reload
```

Open http://localhost:8080/ and see CWL test tools (along with all Galaxy test tools) in left hand tool panel.

To go a step further and actually run CWL jobs within their designated Docker containers, copy the following minimal Galaxy job configuration file to `config/job_conf.xml`. (Adjust the `docker_sudo` parameter based on how you execute Docker).

https://gist.github.com/jmchilton/3997fa471d1b4c556966

Run API tests demonstrating the various CWL demo tools with the following command.

```
./run_tests.sh -api test/api/test_tools_cwl.py
```
## Issues

Work remaining on CWL support for Galaxy is tracked at https://github.com/common-workflow-language/galaxy/issues.

jmchilton committed:
Implement framework for model tools.

This special class of tools leverages the infrastructure for tool inputs, tool state tracking, tool module for workflows, tool API, etc... without actually producing command-line jobs. Instead these tools are provided the input model objects and are expected to produce output model objects directly. This provides an oppertunity

The first driving use case for these tools are also included - namely tools that allow zipping and unzipping paired collections. These tools can be mapped over lists (e.g. list:paired to (list, list) or the inverse) using much of the existing infrastructure for tools. Test cases included that validate these work with mapping operations and in workflows.

The most obvious advantage of these versus traditional tools that do the same thing is that the data isn't copied on disk - new HDAs are created directly from the source datasets.

Testing:

This PR includes various API test cases for functionality, these can be run with the following command:

```
./run_tests.sh -api test/api/test_tools.py:ToolsTestCase.test_unzip_collection
./run_tests.sh -api test/api/test_tools.py:ToolsTestCase.test_zip_inputs
./run_tests.sh -api test/api/test_tools.py:ToolsTestCase.test_zip_list_inputs
./run_tests.sh -api test/api/test_workflows.py:WorkflowsApiTestCase.test_workflow_run_zip_collections
```

jmchilton commented:
This will be rebased frequently but this standing PR should represent the current progress of the cwl branch. Any help would be appreciated, will work on building a TODO list of what needs to be done.

jmchilton committed:
[WIP] Dynamic Galaxy tools.

 - Introduce models and a API for creating tools dynamically.
 - Use Galaxy's testing-only YAML based representation of tools to prototype this.
 - Extend Format 2 workflow definitions to allow embedding tools directly into workflows, either directly or using a CWL-style @import syntax.

Testing:

Test cases demonstrating tools can be imported (only by admins) and are runnable are included with this commit. More test cases regarding workflow use of dynamic tools and Format 2 workflow definition extensions are also included.

These tests can be run with the following commands:

```
./run_tests.sh -api test/api/test_tools.py:ToolsTestCase.test_nonadmin_users_cannot_create_tools
./run_tests.sh -api test/api/test_tools.py:ToolsTestCase.test_dynamic_tool_1
./run_tests.sh -api test/api/test_workflows.py:WorkflowsApiTestCase.test_import_export_dynamic
./run_tests.sh -api test/api/test_workflows_from_yaml.py:WorkflowsFromYamlApiTestCase.test_workflow_embed_tool
./run_tests.sh -api test/api/test_workflows_from_yaml.py:WorkflowsFromYamlApiTestCase.test_workflow_import_tool
```

jmchilton committed:
[WIP] Implement Expression tools in Galaxy.

- Tool definition languge and plumbing and datatype for expressing expressions as jobs.
- Allow connecting expression tools to parameters in workflows, will delay evaluation of workflow so calculated value
- Example test expression tools for testing and demonstration.
- [WIP] Workflow expression module to allow users to specify arbitrary expressions.

jmchilton committed:
Library methods for dealing with sandboxed JavaScript expressions.

Based on work by Peter Amstutz in cwltool (https://github.com/common-workflow-language/cwltool).

jmchilton committed:
New model tool that filters failed datasets out of a collection.

This differs from a traditional tool in that its inputs don't need to be in an 'ok' state and instead of creating new datasets and duplicating data on disk, new HDAs are created from the existing datasets.

jmchilton committed:
New model tool that filters datasets on user supplied expression.

The user is prompted for a JavaScript expression, which is in turn ran once per dataset in a list and used as filter. If the JavaScript evaluates to a Python truthy value, the HDA is copied into the output dataset (without duplicating the data on disk).

The JavaScript expression is supplied various HDA attributes in the environment (currently all metadata values, file_size, file_ext, and dbkey). The supplied test case filters out datasets that do not contain an even number of lines.

Testing:

```
./run_tests.sh -api test/api/test_tools.py:ToolsTestCase.test_filter_0
```

jmchilton committed:
New model tool that flattens any collection into a list.

Testing:

```
./run_tests.sh -framework -id __FLATTEN__
```

jmchilton committed:
Check for nodejs and display useful log message if missing...

... for loading tool actions that require it.

jmchilton committed:
Refactor Filter tool toward more tools that allow expressions.

jmchilton committed:
New model tool that groups datasets in a collection.

Takes in a list dataset collection and produces a list of lists keying the outer list on a user supplied function. This reuses the JavaScript expression code used by the filter model tool.

Testing:

```
./run_tests.sh -framework -id __GROUP__
```

mr-c commented:
Re: "2) For finer grain control within Galaxy, extensions to CWL should allow setting actual Galaxy output types on outputs. (Distinction between fastq and fastqsanger in Galaxy is very important for instance.)"

There should be a corresponding EDAM data format for every galaxy type. For example, http://bioportal.bioontology.org/ontologies/EDAM/?p=classes&conceptid=http%3A%2F%2Fedamontology.org%2Fformat_1932 for Galaxy's "fastqsanger".

It shouldn't require any Galaxy specific extensions to the CWL standard nor tool descriptions; just proper use of what we have already :-)

jmchilton committed:
Reduce duplication in galaxy.tools.parameters.wrapped.

jmchilton committed:
Allow data parameters to specify optional defaults.

A lot more could be done here - support for multiple="true" parameters, collections, testing with identifiers, better error reporting if DatasetWrapper methods are used. I'll follow up with those things if people find the current approach acceptable.

Implements #1682.

pcm32 committed:
Initial commit for collaboration on the k8s runner/job configuration.

pcm32 committed:
Adds imports for pykube

pcm32 committed:
Changes import to handle missing dependency

pcm32 committed:
Sample config file for k8s based execution.

pcm32 committed:
First trial at the __init__

pcm32 committed:
Configures pykube to use the kubernetes config file path set in the job_conf.xml, imports Job instead of Pod.

pcm32 committed:
Adds persistent volume claim name example in config.

pcm32 committed:
Initial commit of queue_job plus required helper methods (not finished).

pcm32 committed:
Adds pieces for PersistentVolume and PersistentVolumesClaims

pcm32 committed:
Adds namespace and port proposal for job_conf file.

pcm32 committed:
Minor formatting

pcm32 committed:
Job name and correct spec.

pcm32 committed:
Adds async jobs to monitor queue

pcm32 committed:
Improves job name, prepending string to make it URLable as required by kubernetes

pcm32 committed:
Comments ports usage (we leave this feature for later) and other minor formatting.

pcm32 committed:
changes way that 'command' option is passed to k8s on the Job object to cater for multi line commands.

pcm32 committed:
Implements stop job.

pcm32 committed:
Minor correction of message.

pcm32 committed:
Adds some todo's

pcm32 committed:
Implements watch_item and replaces the watched_items, to use the parent logic.

pcm32 committed:
Changes stop job to scale job to zero replicas instead of deleting the job (for logging purposes)

pcm32 committed:
Minor documentation improvement.

pcm32 committed:
Adds pods retrials for per installation and individual docker destination.

pcm32 committed:
Removes unnecessary __get_job_states

pcm32 committed:
Uses k8s pod log as stdout/stderr.

pcm32 committed:
Removes unnecessary get CLI plugins method.

pcm32 committed:
PEP8 compliant formatting

pcm32 committed:
Clean up for queue method.

pcm32 committed:
Removes apparently unnecessary parse_destination_params...

pcm32 committed:
Removes url_to_destination, as this is only for legacy destinations (which I guess our's isn't, wouldn't have any URL to transform either)

pcm32 committed:
Adds some logging plus job state changes.

pcm32 committed:
Fixes initial handling of runner params.

pcm32 committed:
Starts monitor thread and  worker threads.

pcm32 committed:
Changes k8s job api path to extensions/v1beta1 as pykube doesn't support the other one yet.

pcm32 committed:
Fixes labels and faulty spec hierarchy.

pcm32 committed:
Fixes volumes and containers to be as expected json arrays.

pcm32 committed:
Fixes non-DNS persistent volume claim.

pcm32 committed:
Improves formatting (PEP suggestions).

pcm32 committed:
Handles correctly error and output files.

pcm32 committed:
Removes some unneeded imports.

davebx committed:
Move metadata-related methods from shed_util_common to metadata_util.

davebx committed:
Correct travis errors and missing imports.

bgruening committed:
Merge branch 'release_16.04' of https://github.com/galaxyproject/galaxy into release_16.04

pcm32 committed:
Missing os environment functionality import (forgot to include in commit).

pcm32 committed:
Merge pull request #1 from galaxyproject/dev

Staying up to date with galaxy dev

pcm32 committed:
Merge branch 'dev' into feature/k8s_job_runner

# Conflicts:
#	.gitignore

pcm32 committed:
Adds job_conf documentation for Kubernetes runner on the job_conf.xml.sample_advanced file.

davebx committed:
Update method call.

davebx committed:
Fix category/repositories API endpoint. Add test.

davebx committed:
Placate Travis.

pcm32 committed:
Describes scenario with <interpreter> <script> inside assumed working directories in containers.

pcm32 committed:
Removes job_conf_k8s.xml file

pcm32 committed:
Adds temporary dependency on my version of pykube, until changes on pykube are pulled and a new release is available.

pcm32 committed:
Fixes typo on comment of runner documentation.

pcm32 committed:
Fixes other errors spotted by travis.

davebx committed:
Refactor out repository-related methods from shed_util_common.

davebx committed:
Move more methods from shed_util_common to repository_util.

pcm32 committed:
Small typos/issues fixed for travis.

davebx committed:
Move a lot more things from shed_util_common to repository_util.

davebx committed:
Fix travis errors.

davebx committed:
Remove another unused import.

jmchilton committed:
Implement a subset of the Common Workflow Language Draft 3 tool format.

CWL Support:
--------------

 - Implemented integer, long, float, double, boolean, and File parameters, and arrays thereof as well some simple unions of these parameters and Any-type parameters. More complex unions of datatypes are stil unsupported.
 - Draft 3 ``CreateFileRequirement``s are supported (see the ``test_rename`` test case).
 - Draft 3 ``InlineJavascriptRequirement`` are support to define output files (see ``test_cat3`` test case).
 - ``EnvVarRequirement``s are supported (see the ``test_env_tool1`` and ``test_env_tool2`` test cases).
 - Secondary files are supported at least partially, see the ``index1`` and ``showindex1`` CWL tools created to verify this as well as the ``test_index1`` test case.
 - Docker integration is only partial (simple docker pull is supported) - so ``cat3-tool.cwl`` works for example. Full semantics of CWL docker support has yet to be implemented. The remaining work is straight-forward and trackd in the meta-issue https://github.com/galaxyproject/galaxy/issues/1684.
 - Expression tools are supported (see ``parseInt-tool`` test case).
 - Non-File CWL outputs are represented as ``expression.json`` files. Traditionally Galaxy hasn't supported non-File outputs from tools but CWL Galaxy has work in progress on bringing native Galaxy support for such outputs https://github.com/common-workflow-language/galaxy/issues/27.

Implementation Notes:
----------------------

 - CWL secondary files are stored in ``__secondary_files__`` directory in the dataset's extra_files_path directory.
 - The tool execution API has been extended to add a ``inputs_representation`` parameter that can be set to "cwl" now. The ``cwl`` representation for running tools corresonding to the CWL job json format with {class: "File: path: "/path/to/file"} inputs replaced with {"src": "hda", "id": "<dataset_id>"}. Code for building these requests for CWL job json is available in the test class.
 - Since the CWL <-> Galaxy parameter translation may change over time, for instance if Galaxy develops or refines parameter classes - CWL state and CWL state version is tracked in the database and hopefully for reruns, etc... we could update the Galaxy state from an older version to a new one.
 - CWL allows output parameters to be either ``File`` or non-``File`` and determined at runtime, so ``galaxy.json`` is used to dynamically adjust output extension as needed for non-``File`` parameters.

Implementation Description:
-----------------------------

The reference implementation Python library (mainly developed by Peter Amstutz - https://github.com/common-workflow-language/common-workflow-language/tree/master/reference) is used to load tool files ending with ``.json`` or ``.cwl`` and proxy objects are created to adapt these tools to Galaxy representations. In particular input and output descriptions are loaded from the tool.

When the tool is submitted, a special specialized tool class is used to build a cwltool compatible job description from the supplied Galaxy inputs and the CWL reference implementation is used to generate a CWL reference implementation Job object. A command-line is generated from this Job object.

As a result of this - Galaxy largely does not need to worry about the details of command-line adapters, expressions, etc....

Galaxy writes a description of the CWL job that it can reload to the job working directory. After the process is complete (on the Galaxy compute server, but outside the Docker container) this representation is reloaded and the dynamic outputs are discovered and moved to fixed locations as expected by Galaxy. CWL allows for much more expressive output locations than Galaxy, for better or worse, and this step uses cwltool to adapt CWL to Galaxy outputs.

Currently all ``File`` outputs are sniffed to determined a Galaxy datatype, CWL draft 3 allows refinement on this and this remains work to be done.

  1) CWL should support EDAM declaration of types and Galaxy should provide a mapping to core datasets to skip sniffing is types are found.
  2) For finer grain control within Galaxy, extensions to CWL should allow setting actual Galaxy output types on outputs. (Distinction between fastq and fastqsanger in Galaxy is very important for instance.)

Testing:
---------------------

    % git clone https://github.com/common-workflow-language/galaxy.git
    % git checkout cwl
    % cd galaxy
    % virtualenv .venv
    % . .venv/bin/activate
    % pip install cwltool

Start Galaxy.

    % GALAXY_RUN_WITH_TEST_TOOLS=1 run.sh --reload

Open http://localhost:8080/ and see CWL test tools (along with all Galaxy test tools) in left hand tool panel.

To go a step further and actually run CWL jobs within their designated Docker containers, copy the following minimal Galaxy job configuration file to ``config/job_conf.xml``. (Adjust the ``docker_sudo`` parameter based on how you execute Docker).

https://gist.github.com/jmchilton/3997fa471d1b4c556966

Run API tests demonstrating the various CWL demo tools with the following command.

```
./run_tests.sh -api test/api/test_tools_cwl.py
```

Issues
---------------------------------

Work remaining on CWL support for Galaxy is tracked at https://github.com/common-workflow-language/galaxy/issues.

Refactor toward workflow support.

pcm32 committed:
Adheres to pykube new logs method instead of previous get_logs

davebx committed:
Merge remote-tracking branch 'upstream/dev' into refactor_suc

davebx committed:
Remove merge conflict aftereffects.

davebx committed:
Fix another incorrectly resolved merge conflict.

davebx committed:
Fix missing import and unused import.

guerler committed:
Replace frame dictionary with regular view

pcm32 committed:
Merge pull request #2 from galaxyproject/dev

Stays up-to-date with galaxy dev branch

pcm32 committed:
Merge branch 'dev' into feature/k8s_job_runner

jmchilton committed:
Provide backward-compatible galaxy.datatypes.checkers module.

xref http://dev.list.galaxyproject.org/STACKS-error-td4669428.html

pcm32 committed:
Adds the ability to build container ids for pieces for the centralised galaxy containers class and takes into account the default and override modes, suggested by jmchilton.

pcm32 committed:
Moves the kubernetes runner to use the centralised galaxy way of picking containers, allowing destinations to override a tool, use a tool, or have a default container if the tool doesn't provide one.

pcm32 committed:
Additional new required parameter for destinations (that will be used with the kubernetes runner) to play nicely with existing galaxy handling of containers.

pcm32 committed:
Removes quotes to docker_enabled paramter value in documentation of destination associated with Kubernetes.

pcm32 committed:
Adds missing parenthesis for string formatting.

guerler committed:
Avoid empty strings when converting values to display text

nsoranzo committed:
SLURM: when scontrol fails, try to get job state with sacct

pcm32 committed:
Fixes use of container object returned by underlying galaxy methods.

pcm32 committed:
Introduces additional parameter to avoid having the command-line modified for kubernetes.

guerler committed:
Remove grid lock feature

guerler committed:
Use regular initialization from frame shadow view

guerler committed:
Add support for optional header items/icons

guerler committed:
Remove unused class variables, reduce indention

guerler committed:
Parse view instead of content dom on click

guerler committed:
Add event handlers to navigate through history

shiltemann committed:
add mothur.tre datatype (subclass)

shiltemann committed:
scan entire file when setting metadata

for large group files this would not find all groups

bgruening committed:
add is_target_available() to conda utils

bgruening committed:
stdin and stdout needs a PIPE to retrive the stdout

guerler committed:
Cache histories individually

bgruening committed:
Fix bad merge.

nsoranzo committed:
Merge _complete_terminal_job() from BaseJobRunner into DRMAAJobRunner

nsoranzo committed:
Clarify _complete_terminal_job() return value

carlfeberhard committed:
UI, history: correctly handle index API defaults

Bypasses the histories.index default `deleted=False` flag when using
HistoryCollection to fetch with `includeDeleted`.

martenson committed:
Merge pull request #2395 from bgruening/conda_search

Add conda search function

dannon committed:
Merge pull request #2413 from carlfeberhard/fix.history-collection

UI, history: correctly handle index API defaults

jmchilton committed:
Update galaxy-lib related modules to latest in that project.

carlfeberhard committed:
Fix to 4a61a4c: add hids for dataset inputs

guerler committed:
Updates and fixes of the run workflow ui elements and views

guerler committed:
Merge branch 'dev' into scratchbook_slider_000

jmchilton committed:
Remove HTML_CHECK_LINES from checkers export.

As suggested by @nsoranzo and @dannon, we will restore if and only if someone finds a datatype that uses it.

dannon committed:
Merge pull request #2389 from jmchilton/checkers_restore

[16.04] Provide backward compatible galaxy.datatypes.checkers module...

dannon committed:
Merge remote-tracking branch 'upstream/release_16.04' into dev

bgruening committed:
Merge pull request #2396 from guerler/scratchbook_slider_000

Add history navigation options for dataset frames to scratch book

nsoranzo committed:
Raise Exception if watch_tools has an unrecognized value

Also do not log message "Watchdog library unavailable" if watch_tools is
False

jmchilton committed:
Merge pull request #2418 from nsoranzo/toolbox_watcher

Raise Exception if watch_tools has an unrecognized value

nsoranzo committed:
Python3: use log.warning() instead of deprecated log.warn()

$ python3 -Wd -c 'import logging; logging.warn("foo")'
-c:1: DeprecationWarning: The 'warn' function is deprecated, use 'warning' instead
WARNING:root:foo

martenson committed:
wrap bare str raises with generic Exception()
as Pyton 2.7 does not support bare str

martenson committed:
wrap bare str raises with generic Exception()
as Pyton 2.7 does not support bare str

martenson committed:
fix the mess made with mass edit

martenson committed:
fix the mess made with mass edit

martenson committed:
Merge pull request #2414 from jmchilton/lib-fix

Update galaxy-lib related modules to latest in that project.

martenson committed:
Merge pull request #2420 from nsoranzo/log_warn

Python3: use log.warning() instead of deprecated log.warn()

dannon committed:
Merge pull request #2421 from martenson/wrap-str-raises

wrap bare str raises with generic Exception()

dannon committed:
Merge pull request #2415 from carlfeberhard/fix.show-params-hid

Fix to 4a61a4c: add hids for dataset inputs

jmchilton committed:
Merge pull request #2424 from martenson/port-exceptionwrap-16.04

[16.04] wrap bare str exceptions

shiltemann committed:
add otulabels metadata to mothur.otu datatype

martenson committed:
minor docs touch

dannon committed:
Merge pull request #2427 from martenson/minor-docs

minor docs touch

jmchilton committed:
Merge pull request #2416 from guerler/tool_composite_000

Updates and fixes for the run workflow ui elements

hmenager committed:
add and/or correct missing EDAM formats/data to datatypes.

hmenager committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into dev

bgruening committed:
Merge pull request #2428 from hmenager/dev

Correct various EDAM data/format annotations in datatypes

ThomasWollmann committed:
fixed $defaults according to #2430

shiltemann committed:
minimize number of lines read in set_meta functions

natefoo committed:
Only scan the last 16 MB of a job's stderr for slurm memory warning
messages.

natefoo committed:
Fix linting error

dannon committed:
Merge pull request #2432 from natefoo/job-output-memory-fix

[16.04] Only scan the last 16 MB of a job's stderr for slurm memory warning messages

jmchilton committed:
Update Galaxy testing to target latest galaxy-lib.

Incorporates https://github.com/galaxyproject/galaxy-lib/pull/10 which modifies the interface to the test output verification code to allow reuse from within Planemo in both non-tool and non-Galaxy contexts. I also think this represents a better separation of concerns.

One upshot of this is that CWL-style checksum verification is now possible from within Galaxy tools - a test case of this is included.

guerler committed:
Properly display subworkflows in new run workflow form

guerler committed:
Remove output details from success message

nsoranzo committed:
Merge pull request #2436 from jmchilton/galaxy_lib_testing_update

Update Galaxy testing to target latest galaxy-lib.

pcm32 committed:
Merge pull request #3 from galaxyproject/dev

Updates to current galaxy dev

pcm32 committed:
Merge branch 'dev' into feature/k8s_job_runner

guerler committed:
Re-add new history links to success message

martenson committed:
isolate the data libraries css to affect only the libraries app

ThomasWollmann committed:
polish

nsoranzo committed:
Log SLURM accounting storage error

shiltemann committed:
update superclass accessing

carlfeberhard committed:
Raven: fix import in galaxy.panels.mako

Raven now uses a UMD and checks for requirejs removing it from globals
if found. This adds a require import to load Raven.

carlfeberhard committed:
Raven: revert previous commit and move raven first

Moves galaxy.panels.mako's import of Raven back to vanilla script
tag and before the requirejs library import, allowing Raven's UMD
to define Raven as a global.

This starts Raven/sentry *before* the library package is loaded so that
library errors can be captured as well.

carlfeberhard committed:
Raven: remove tracekit as raven has its own

nsoranzo committed:
Overwrite the correct part of stderr

- Skip partial line
- Seek back to the correct position after readlines()
- Use the same approach in __check_memory_limit()

martenson committed:
Merge pull request #2445 from carlfeberhard/fix.raven

Fix raven.js import order in galaxy.panels.mako

dannon committed:
Merge pull request #2447 from nsoranzo/release_16.04_fix_for_2432

[16.04] Overwrite the correct part of stderr

guerler committed:
Fix batchmode labels

nsoranzo committed:
Merge branch 'release_16.04' into dev

jmchilton committed:
Merge pull request #2440 from guerler/fix_subworkflows_display

Properly list subworkflows in run workflow form

bgruening committed:
Merge pull request #2438 from ThomasWollmann/dev

fixed $defaults according to #2430

bgruening committed:
Merge pull request #2429 from shiltemann/mothur_datatypes

small updates to Mothur datatypes

leobiscassi committed:
Add support to pdbqt datatype

davebx committed:
Create categories in the right order.

leobiscassi committed:
Adjust set_peek method of the pdbqt class

leobiscassi committed:
fix pep8 errors

nsoranzo committed:
Make some files compatible with Python3

Also enlarge the set of files tested with flake8 under Python3

leobiscassi committed:
Adjust sniff function

bgruening committed:
Merge pull request #2452 from leobiscassi/add_pdbqt_datatype

Add support to PDBQT molecule datatype

jmchilton committed:
Merge pull request #2444 from martenson/isolate-lib-css

isolate the data libraries css to affect only the libraries app

guerler committed:
Relax validation condition

jmchilton committed:
Merge pull request #2314 from phnmnl/feature/k8s_job_runner

Kubernetes job runner

guerler committed:
Revise contradicting test case, workflow injection through api attempts full validation without available datasets

jmchilton committed:
Merge pull request #2403 from nsoranzo/slurm

SLURM: when scontrol fails, try to get job state with sacct

bgruening committed:
Fix usage of super() in the mothur datatypes.

martenson committed:
Merge pull request #2457 from bgruening/mothur_fix

Fix usage of super() in the mothur datatypes.

martenson committed:
Merge pull request #2453 from nsoranzo/python3

Python3: Make some files compatible with Python3

guerler committed:
Add break to interrupt loop

guerler committed:
Filter missing inputs when processing data

blankenberg committed:
Do not encode dict items that start with 'metadata_' in return value of api/tools/create.

Resolves #2423, #2137, and https://github.com/galaxyproject/tools-iuc/pull/770. However, this whole act of encoding by key.endswith('_id') is very messy.

blankenberg committed:
Add test for failing handling of BIOM1 metadata.

martenson committed:
Merge pull request #2458 from guerler/fix_process_data

[16.04] Filter missing inputs when processing data

jmchilton committed:
Merge pull request #2459 from blankenberg/metadata-id-fix

[16.04] Do not encode dict items that start with 'metadata_' in return value of api/tools/create.

guerler committed:
Add parameter expansion caller and test cases as preparation for workflow api extension

martenson committed:
Merge pull request #2269 from davebx/refactor_suc

Move specific methods from shed_util_common.py to specific_util.py

nsoranzo committed:
Delay import of DrmaaSessionFactory after runner initialization

Fix issue #2463, i.e. the regression introduced in #2102 by which Galaxy
no longer respects <param id="drmaa_library_path"> in config/job_conf.xml ,
preventing the use of multiple drmaa runners.

martenson committed:
Merge pull request #2466 from nsoranzo/release_16.04_fix_for_2463

[16.04] Delay import of DrmaaSessionFactory after runner initialization

nsoranzo committed:
Merge branch 'release_16.04' into dev

jmchilton committed:
Make pykube an optional dependency since we don't have wheels for it.

I should have requested this as a change to https://github.com/galaxyproject/galaxy/pull/2314.

Rebased with improved job conf sample comment thanks to suggestions by @nsoranzo.

nsoranzo committed:
Fix py34-lint tox test

Broken with the merge of #2269.

martenson committed:
Merge pull request #2469 from nsoranzo/python3

Fix py34-lint tox test

martenson committed:
Merge pull request #2468 from jmchilton/pykube_optional

Make pykube an optional dependency since we don't have wheels for it.

jmchilton committed:
Merge pull request #2464 from guerler/add_workflow_parameter_expansion_caller

Add parameter expansion caller and test cases as preparation for workflow api extension

guerler committed:
Prevent missing display to text values in workflow preview

nsoranzo committed:
Re-add get_updated_changeset_revisions_from_tool_shed() method

It was accidentally removed in PR #2269.
Fix toolshed tests

martenson committed:
Merge pull request #2471 from guerler/fix_workflow_display

[16.04] Prevent missing display to text values in workflow preview

nsoranzo committed:
Merge pull request #2473 from martenson/port-value-to-display

[16.04] Avoid empty strings when converting values to display text

bgruening committed:
Merge branch 'release_16.04' of https://github.com/galaxyproject/galaxy into release_16.04

nsoranzo committed:
Merge branch 'release_16.04' into dev

martenson committed:
Merge pull request #2472 from nsoranzo/fix_for_2269

Re-add get_updated_changeset_revisions_from_tool_shed() method

davebx committed:
Workaround for circular import situation.

davebx committed:
Move directory_hash_id into galaxy/util.py. Use util.directory_hash_id throughout.

davebx committed:
Remove now-unused import.

jmchilton committed:
Merge pull request #2477 from davebx/circular_workaround

Workaround for circular import. Refactor.

jmchilton committed:
Set a default for tool_dependencies_dir.

This should make some obsecure errors less easy to encounter and makes it less work to get going with conda or the Tool Shed.

With config sample fix from @martenson.

martenson committed:
fix markup for button links on sharing template
(ff would fail to generate images)
turn of autocomplete

bgruening committed:
Merge pull request #2480 from jmchilton/tool_dependencies_default

Set a default for tool_dependencies_dir.

nsoranzo committed:
Fix for job handlers losing track of jobs when using DRMAA with internalexception_retries or invalidjobexception_retries

nsoranzo committed:
Log all failed job status checks

shiltemann committed:
don't generate error for pairwise distance matrices

martenson committed:
Merge pull request #2484 from nsoranzo/release_16.01_fix_for_4a024b882f504aded1db32f1ed056f7c9109d218

[16.01] Fix for job handlers losing track of jobs when using DRMAA with internalexception_retries or invalidjobexception_retries

nsoranzo committed:
Merge branch 'release_16.01' into release_16.04

Conflicts:
	lib/galaxy/jobs/runners/drmaa.py

nsoranzo committed:
Merge branch 'release_16.04' into dev

nsoranzo committed:
Reset exception retries after successful job status check

bgruening committed:
Merge pull request #2481 from martenson/fix-html-sharing

[16.04] fix markup for button links on sharing template

martenson committed:
fix image generation for workflows with missing tools
currently this ends up with a server error

martenson committed:
drop unnecessary raise and fix capitalization
incorporate comments of Bjoern and Nicola

martenson committed:
Merge pull request #2456 from guerler/fix_building_mode

[16.04] Relax validation condition for dynamic parameters

nsoranzo committed:
Merge pull request #2488 from martenson/wrap-svg-for-missing-tools

[16.04] fix image generation for workflows with missing tools

bgruening committed:
Merge branch 'release_16.04' of https://github.com/galaxyproject/galaxy into release_16.04

bgruening committed:
add IUC channel to the list

dannon committed:
Merge pull request #2490 from bgruening/iuc_conda_channel

[16.04] Add IUC conda channel

bgruening committed:
Merge pull request #2489 from shiltemann/dev

mothur datatypes: don't generate error for pairwise distance matrices

martenson committed:
Merge pull request #2487 from nsoranzo/release_16.04_reset_exception_retries

[16.04] Reset exception retries after successful job status check

nsoranzo committed:
Merge branch 'release_16.04' into dev

Conflicts:
	test/unit/tools/test_select_parameters.py

jmchilton committed:
Update cwl processing for recent changes to cwltool and galaxy.

jmchilton commented:
This PR has be superseded by #47 - work will now be tracked in the cwl-1.0 branch.
